### ISSUE TITLE
Discussão Odin v3.0.0

### ISSUE THREAD

**Issue by adammacias** [wpbrasil/odin#328]
Pessoal, o @claudiosmweb sugeriu um hangout para discutirmos mudanças radicais no Odin, estou abrindo esse tópico para receber feedbacks do que podemos criar/remover/mudar e etc, daí a gente faz o hangout com base nesse tópico! :dancer: 

cc:
@wpbrasil/owners 
@wpbrasil/developers 
@wpbrasil/contribuidores

**Comment by AdsonCicilioti** [comment]
Bacana! Gostaria de colaborar com um redesign no BRANDING do Odin. Acho que a v3.0 merece. Mas tbm tive umas idéias q vou tentar reunir.

**Comment by matheusgimenez** [comment]
Com certeza estarei presente!

**Comment by nandomoreirame** [comment]
Legal, vou acompanhar com certeza..

Uma dia: Talvez fazer do Odin menos tema e mais framework... mais modularizado... tem muitas pessoas que não usam por achar que se trata de "mais um tema" no mercado...

Abrax

**Comment by rickbenetti** [comment]
Minha sugestão pode parecer contra a maré, mas eu tiraria o Bootstrap ou a necessidade dele, ter um padrão de classes próprio o tornaria mais universal.

**Comment by matheusgimenez** [comment]
Acho importante levar em conta o issue #301, acho importante tentarmos manter no padrão do WP e deixarmos de lado a Settings API :)

**Comment by valeriosouza** [comment]
Não concordo em tirar o bootstrap e sim de facilitar a forma de incluir outros.

**Comment by AdsonCicilioti** [comment]
Verdade @valeriosouza. Inclusive facilitar a inclusão de temas do próprio bootstrap, como os do https://bootswatch.com/

**Comment by AdsonCicilioti** [comment]
Outra feature que poderia ser implementada é a customização dos mapas do google utilizando APi style dos mapas ou até mesmo algo mais elaborado com no https://snazzymaps.com/.

**Comment by adammacias** [comment]
@valeriosouza concordo, acho que talvez uma boa solução para isso, seria deixar o Odin 100% preparado para ser um tema pai, e criar um tema filho o mais limpo possível, com o Bootstrap, Foundation ou etc.

**Comment by matheusgimenez** [comment]
Não gosto da ideia de deixar o Odin como tema pai e filho ser editavel.. Eu por exemplo (com a Brasa), uso temas filhos de temas feito com o Odin em alguns projetos e com isso ficaria impossivel, quer dizer, não dá pra criar um tema neto..

**Comment by rickbenetti** [comment]
Concordo com o @adammacias sobre o Odin ser um tema pai, eu faço isso com todos os projetos eu tenho 1 tema pai e temas filho para cada necessidade.

Mas veja bem @matheusgimenez neste caso eu acredito que deveria ser separado as versões assim com o Bootstrap está fazendo, a versão 2 continuou existindo, a versão 3 continuará, e a versão 4 e 5 surgirão, mas um projeto uma vez feito com uma versão é custoso ficar mudando, então acredito mais em mudança no seu caso para projetos novos e não contínuos.

Posso estar falando asneira, não ajudei na criaçãod o Odin, achei muita coisa bacana e muita coisa ruim, mas acredito que ele precisa agora tomar um rumo de amadurecimento de versões o que tem hoje pode e deve ser melhorado, mas não inviabilizar o que já existe, para mudanças o melhor seria uma branch com versão totalmente nova igual Bootstrap e Foundation fazem.

**Comment by allysonsouza** [comment]
Não consigo ver benefícios em pensar o Odin como um tema preparado para ser um "parent theme". 

Como ficaria o projeto? Viria com um Odin Theme + um My Child Theme com os templates duplicados? Ou fica por conta do usuário, sobrescrever os templates que deseja? No caso, eu sobrescreveria todos, porquê o Odin é um framework e a maioria dos templates dele precisam ser trabalhados para atender o design dos projetos... Acho uma abordagem estranha pelo fato do Odin não ser de fato um tema que necessitará de poucas alterações.

**Comment by diegorojas** [comment]
Ola gente, tudo bem? Como temos usado muito o Odin entendo esse horizonte de transforma-lo em um Parent Theme, mas a questão é que ele não foi concebido para isso. Ele foi feito para ser um framework gerador de Temas e isso não pode mudar, ao meu ver.

Temos dois caminhos... Fazer um projeto filho do Odin, que seria o Thor (por exemplo) para esse sim ser um Parent Theme, mas que precisaríamos pensar na forma de manter os dois projetos atualizados de forma integrada e simples..

E o outro caminho, mais legal ao meu ver, é de pensar em um interface front-end geradora de temas, mais ou menos como o _s faz, mas mais complexa.. O usuário preenche uns formularios, seta umas coisas, coloca os nomes de CPT que vai querer a a interfcace "cospe" do Odin um Parent Theme... Isso seria muito tesão... Mas é um projeto razoavelmente trabalhoso... o que acham?

**Comment by claudiosmweb** [comment]
Podem esquecer de ele virar um Parent Theme. O que pode acontecer é retirarmos várias coisas da versão 3.0.0 e transformar em plugins, mas não vai virar Parent Theme.

**Comment by AdsonCicilioti** [comment]
O Odin já é o pai de todos. :+1:

**Comment by claudiosmweb** [comment]
Sim, ideia é criar seu tema por cima dele e pronto.
Não manter ele como tema pai, nem vale esse esforço, além que hoje como ele está também nem é complicado fazer isso, já tem várias coisas que torna isso possível.
Então forcar nisso é cagada.

**Comment by AdsonCicilioti** [comment]
@claudiosmweb aproveitando que vc está na área, estive observando a feature do tema Layers que torna diversas áreas personalizáveis, meio que 'widgetizando' elas para facilitar a edição/customização dos conteúdos no front através do "Personalizar" padrão do WP.
Como vc vê isso? Seria viável a implementação neste sentido no Odin?

**Comment by claudiosmweb** [comment]
@AdsonCicilioti não, ideia aqui não é arrasta e solta.

**Comment by AdsonCicilioti** [comment]
Certo. E o que vc já tem como meta para a v3.0?

**Comment by claudiosmweb** [comment]
@AdsonCicilioti melhorar as classes do plugin, tornar mais simples ainda de criar as coisas com ele.
Rever algumas funcionalidades e remover o que não é usado.
Tudo coisa tangível, nada surreal ou que galera só vai dar ideia, não ter noção de como fazer e muito menos querer ajudar a desenvolver. O que eu quero qualquer um consegue ajudar.

**Comment by AdsonCicilioti** [comment]
Sim claro. Verdade. Idéias vão chover mesmo. Queria saber se o Odin iria seguir essa linha de um tema para "não-programadores", o tornando mais digerível pro mercado, mas isso ficará a cargo de como vamos utilizá-lo em nossos projetos.

**Comment by claudiosmweb** [comment]
@AdsonCicilioti me passa essas drogas ai que quero usar também.
Não foi e nunca será **"não-programadores"**, foco aqui sempre foi desenvolvimento rápido e ágil, para entregar algo limpo e simples para o cliente final. Cliente final vai fazer apenas o que foi deixado de opção do tema pra ele depois e ponto.
Ideia aqui nunca foi ser um Visual Composer, até porque se quer arrastar e soltar coisas que vá usar ele e não o Odin.

**Comment by AdsonCicilioti** [comment]
Blz brow. Foi o que eu disse. Tais implementações, caso necessário, ficará a cargo de cada projeto, via plugin ou implementações 'na unha'.
Valew meu querido! Tamo junto!

Quanto as drogas, desculpa.. vou ficar te devendo brow...  Eu viajo muito mesmo.. imaginação muito fértil! XD

**Comment by leobaiano** [comment]
Nessa próxima versão quero contribuir mais com o Odin. Acho que seria legal fazer uma lista do que entra e o que sai nessa próxima versão. Por exemplo, a criação de CPT e taxonomy são as funcionalidades que mais usei nos projetos com Odin, porém o ideal é criar essas coisas como plugin para que não fique atrelado ao visual, vai manter CPT e taxonomy no Odin? Das coisas para incluir acho que tem que criar uma classe para o customizer (se ainda não existir).

**Comment by matheusgimenez** [comment]
Sobre o customizer estamos debatendo isso no issue #301

**Comment by AdsonCicilioti** [comment]
Quando mencionei o jeitão do tema Layers, me referi a forma em fazer uso melhorado do Customizer, para editar conteúdo e não Layout. Mas o @claudiosmweb entendeu Drag'n-Drop.. Enfim, acho que seria uma boa fazer uso deste recurso já que é padrão e nativo do WP.

Vou tentar colaborar com isso @matheusgimenez.

**Comment by claudiosmweb** [comment]
Mesmo assim, Customize API é algo que provavelmente não vai entrar no Odin, simplesmente porque não precisa. O WordPress já tem uma API excelente para isso, sem a necessidade de ficar lidando como as opções aparecem (HTML e CSS), tanto quanto funcionalidades como no caso de upload ou adicionar colorpicker, sem contar também para salvar.

Alias, uma das ideias é transformar a forma que criamos algumas coisas aqui, iguais aos do Customize API.
Assim todo mundo fica com a mesma roupa.

Então como você falou @AdsonCicilioti: "Enfim, acho que seria uma boa fazer uso deste recurso já que é padrão e nativo do WP.". SIM vai lá e usa.

**Comment by claudiosmweb** [comment]
@leobaiano sobre essas coisas, sim o ideal é criar por plugin, mas aqui desde sempre focamos em entregar o projeto pronto para o cliente e livre dele alterar e quebrar o que foi implementado, o que iria acontecer usando plugins para fazer isso.
No formato que estamos aqui, não seria um problema ter isso dentro do Odin, entretanto se for levar tudo o que tem aqui para um plugin, ai também não seria nenhum problema.

**Comment by allysonsouza** [comment]
Concordo com o @claudiosmweb, estou fazendo um projeto com o Odin e utilizando a API de Customizer e a coisa já tá fina, fazer algo para criar os mesmos campos seria uma redundância, e no caso de precisar de campos adicionais o plugin que o @matheusgimenez postou na issue #301 atende bem.

**Comment by AdsonCicilioti** [comment]
Entendi @allysonsouza. Valew

**Comment by claudiosmweb** [comment]
Hoje temos um problema enorme aqui de repetir código, todas as classes repetem código para gerar HTML das opções, sem contar outras coisas.
Isso é algo que precisamos resolver nessa versão nova.

Até podemos adicionar algumas coisas do Kirki aqui para ajudar na Customize API, mas não fazer um wrapper dela como fazer para outras coisas.

**Comment by adammacias** [comment]
Como o @claudiosmweb disse, tem boilerplate pacas nas classes! :+1: Isso esta dificultando a manutenção.

***

Outra coisa que eu gostaria saber a opinião de vocês, é, vale a pena trocar o **Grunt** pelo **Gulp**? :ghost:

**Comment by claudiosmweb** [comment]
@adammacias acho que vale xD
Sinceramente eu comecei ser totalmente 100% a favor do gulp nas ultimas semanas :p

**Comment by adammacias** [comment]
Show @claudiosmweb , o @fdaciuk tbm curti Gulp que eu sei, nunca tive a oportunidade de usa-lo ainda, mas confio em vocês que seja melhor que Grunt! haha

**Comment by claudiosmweb** [comment]
Mas estou vendo que vou ter que criar uns modulos para o Gulp para fazer umas coisas que quero aqui xD

**Comment by fdaciuk** [comment]
Preciso tomar vergonha na cara e terminar o odin-gulp :stuck_out_tongue_winking_eye:

**Comment by trsenna** [comment]
cai no site do odin agora pouco, e parei aqui depois de navegar um pouco na documentacao e código. Permitam-me opinar sobre o que eu acharia legal numa próxima versao

* que o odin continuasse sendo um tema de partida, e moveria as funcionalidades de framework para um plugin ou outro projeto no github (para ser incluido usando submodule)
* vi que ja foi discutido nas threads anteriores, mas concordo que tem um pouco de codigo repetido, principalmente codigo q gera formularios. Seria legal se encontrassem algum jeito de nao acontecer estas repeticoes.
* uma idéia q tive recentemente para meus projetos em WordPress, foi deixar pronto alguns post-types (e taxonomias tambem) mais utilizados, que preciso apenas fazer um include no functions.php pra utiliza-los, ao inves de recorrer a plugins. Alguns post-types seriam portfolio, testemunhos e etc... eu pelo menos gosto de evitar plugins as vezes.
* a documentação está muito legal mesmo, seria legal manter e documentar mais coisas

Acho que esses pontos mais o q ja foi discutido na thread já fazem justiça a uma versao 3.0 :)

**Comment by matheuscl** [comment]
Acredito que uma feature bacana seria grupo de campos com a funcionalidade de `repeater`, algo como o [ACF](http://www.advancedcustomfields.com/add-ons/repeater-field/).

**Comment by luizbills** [comment]
`repeater` seria muito útil mesmo. :+1:

**Comment by allysonsouza** [comment]
Moçada, fiz esse esquema com uma proposta de como poderíamos alterar a abordagem do Odin para a versão 3.0. Fiz dessa forma mais visual pois acredito que deixa bem claro as ideias. Nessa proposta penso que solucionaríamos várias questões, desde definir ou não CPT's em temas ou plugins, até utilizar ou não o Odin como um framework para plugins também.

O que vocês acham?

![odin schema](https://cloud.githubusercontent.com/assets/2000638/14209855/949b206a-f7fc-11e5-98bd-b11a3b8c46fe.png)

**Comment by allysonsouza** [comment]
No caso de utilizar os recursos do Odin Toolkit no tema, seria só fazer como já é feito com WooCommerce, com um 'is_odin_toolkit_activated'.

**Comment by adammacias** [comment]
Alguns não sabem se o **Odin** é um tema base ou um framework, e na verdade, hoje são os dois em um só! :dancers: 

@allysonsouza Gostei da proposta de separar o **tema** do **framework**.  Vai dar mais liberdade para criar outros tipos de temas. Exemplo: Tema voltado para imóveis, outro para portfolio e etc.

Mas como acredito que o Odin é mais famoso mais pelo:

> _**FRAMEWORK** QUE CHUTA BUNDAS._

Eu sugiro manter o nome **Odin** a fim de aplicar a ideia do *Toolkit*. 

Como eles não vão depender um do outro, acho que o nome do tema poderia ser algo como: "Thor Theme", coisa que já tentaram fazer antes e eu gostei na época. Já tem até um repositório criado aqui no wpbrasil, olha https://github.com/wpbrasil/thor

![unnamed0](https://cloud.githubusercontent.com/assets/1139893/14489676/e97dfc1e-0144-11e6-85f8-a836eb2f1565.png)

**Comment by RafaelFunchal** [comment]
Curti também essa ideia de separar o framework do tema. 

Assim ainda podemos esgotar todos os nomes nórdicos :)

**Comment by allysonsouza** [comment]
Curti @adammacias!

@RafaelFunchal Vamos fazer o panetão nórdico inteiro, tema base pra moçada não baixar comprar mais tema algum xD

**Comment by cleberoliveira** [comment]
Acho que um nome bacana para o theme base poderia ser Hugin, fazendo ref. aos corvos informantes de odin :P 

https://pt.wikipedia.org/wiki/Hugin_e_Munin

**Comment by RafaelFunchal** [comment]
Pronto, mais fácil:

https://pt.wikipedia.org/wiki/Lista_de_nomes_da_mitologia_n%C3%B3rdica

**Comment by allysonsouza** [comment]
@cleberoliveira acho que podemos deixar Odin/Thor, que já existem, e nessa nova arquitetura, poderemos ter Hugin's, Munin's, Lokis, Freyas, Siegfrieds... Vai da comunidade se engajar e se envolver no desenvolvimento dos temas :smile:

**Comment by adammacias** [comment]
Né, só temos que ir com calma antes de criar uma lista enorme de temas e depois ficar abandonado/defasado haha vamos por partes amiguinhos. :dancers:

**Comment by allysonsouza** [comment]
Peguei um tempo aqui e fiz uma pequena separação, daquilo que é considerado adequado ao tema, e do que já entreva no [_plugin territory_](https://make.wordpress.org/themes/handbook/review/required/explanations-and-examples/#plugin-territory). (Basicamente eram as classes que possibilitavam a criação de tipos de conteúdo, taxonomias e novas meta boxes para os tipos de conteúdo.)

Utilizei o TGM (http://tgmpluginactivation.com/), método bastante utilizado para definir dependência e recomendação de plugins. A princípio fiz no Github da Haste para não poluir o código aqui/ficar criando branch.

**Odin**
[https://github.com/HasteDesign/odin/tree/v3-dev](https://github.com/HasteDesign/odin/tree/v3-dev)
É o mesmo que você conhece, só não vai ter os recursos de post type, taxonomy, meta boxes, etc.
No functions.php você pode definir que o Odin Toolkit (plugin) é recomendado (ou, configurando mais um pouco, que é obrigatório) para o funcionamento do seu tema.
```
/**
 * TGM.
 */
 //require_once get_template_directory() . '/inc/tgm/odin-required-plugins.php';
```

Para utilizar os recursos que dependem do plugin, poderia ser feito tal como é com o Woocommerce hoje, criando uma função que verifica se o plugin está ativo (a partir da existência da classe Odin_Toolkit).

**Odin Toolkit**
[https://github.com/HasteDesign/odin-toolkit](https://github.com/HasteDesign/odin-toolkit)
Fiz um plugin bem simples, que basicamente carrega as classes, deixando-as disponíveis para serem utilizadas no tema, no próprio plugin ou em outro plugin. Assim o desenvolvedor escolhe: 

1. Quer criar cpt no tema? Ok.

1. Quer criar um plugin específico para os custom post type, taxonomies, meta boxes, mas que dependam do Odin Toolkit? Ok.

1. Quer modificar o Odin Toolkit para ser um plugin personalizado, do qual o tema do seu projeto irá depender? Ok.

Uma questão que fiquei pensando é sobre as classes carregadas pelo plugin. Penso que talvez  elas possam ser carregadas como módulos (estilo Jetpack), oferecendo uma interface no painel para definir o que será utilizado. (Ai seria preciso, para cada recurso que for ser utilizado, verificar se o recurso específico está ativo, como por exemplo criar um custom post type).

Com isso temos outro ponto: tanto o tema quanto o plugin poderiam estar no repositório oficial.

O que vocês acham? É um caminho interessante para todos que estão utilizando o Odin?

**Comment by allysonsouza** [comment]
Bom pessoal, gostaria que se manifestassem sobre o assunto, acho que o Odin está ficando um pouco defasado, limitando a ampliação da sua utilização do modo como está hoje, acho que precisamos definir o futuro do seu direcionamento para melhorá-lo.

**Comment by AdsonCicilioti** [comment]
Bacana @allysonsouza, tem uma pá de coisinhas que precisamos fazer..
- Start Content, que saiu na nova versão do WP 4.7.
- Explorar mais o Customizer...
- Escolher uma Biblioteca CSS mais leve e atual como [Spectre.css](https://github.com/picturepan2/spectre) (Bem completo e apenas 30kb)

Vou dar uma olhada no que mais podemos fazer, mas por hora quero dizer que pode contar comigo.

**Comment by dallington** [comment]
E o Odin morreu ?

**Comment by RafaelFunchal** [comment]
Se puder ajudar a manter vivo, só comemorar.
